subsys/suit/ipuc/src/suit_ipuc_sdfw.c

```diff
@@ -0,0 +1,695 @@
+			if (suit_flash_sink_get(&ipuc_sink, (uint8_t *)slot_address, slot_size) ==
+			    SUIT_PLAT_SUCCESS) {
+				ipuc_sink_allocated = true;
+			}
+		}
+
+		if (!ipuc_sink_allocated) {
+			k_mutex_unlock(&ipuc_mutex);
+			return SUIT_PLAT_ERR_IO;
+		}
+
+#if CONFIG_SUIT_DIGEST_CACHE
+		suit_plat_digest_cache_remove(component_id);
+#endif /* CONFIG_SUIT_DIGEST_CACHE */
+
+		if (ipuc_sink.erase == NULL ||
+		    ipuc_sink.erase(ipuc_sink.ctx) != SUIT_PLAT_SUCCESS) {
+			LOG_ERR("Erase operation not supported or failed");
+			if (ipuc_sink.release) {
+				ipuc_sink.release(ipuc_sink.ctx);
+			}
+			k_mutex_unlock(&ipuc_mutex);
+			return SUIT_PLAT_ERR_IO;
+		}
+
+		if (ipuc_sink.release) {
+			ipuc_sink.release(ipuc_sink.ctx);
+		}
+	}
+
+	ipuc_entry->usage = IPUC_IPC_IN_PLACE_UPDATE;
+	ipuc_entry->ipc_client_id = ipc_client_id;
+	ipuc_entry->write_peek_offset = 0;
+	ipuc_entry->last_chunk_stored = false;
+
+	LOG_INF("suit_ipuc_sdfw_write_setup success, IPUC address: %p", (void *)slot_address);
+
+	k_mutex_unlock(&ipuc_mutex);
+	return SUIT_PLAT_SUCCESS;
+}
+
+suit_plat_err_t suit_ipuc_sdfw_write(int ipc_client_id, struct zcbor_string *component_id,
+				     size_t offset, uintptr_t buffer, size_t chunk_size,
+				     bool last_chunk)
+{
+	intptr_t slot_address = 0;
+	size_t slot_size = 0;
+
+	if (component_id == NULL) {
+		return SUIT_PLAT_ERR_INVAL;
+	}
+
+	if (ipuc_decode_mem_component_id(component_id, &slot_address, &slot_size) !=
+	    SUIT_PLAT_SUCCESS) {
+		/* Only MRAM-based MEM components are supported
+		 */
+		return SUIT_PLAT_ERR_UNSUPPORTED;
+	}
+
+	k_mutex_lock(&ipuc_mutex, K_FOREVER);
+	suit_ipuc_entry_t *ipuc_entry = ipuc_entry_from_component_id(component_id);
+
+	if (ipuc_entry == NULL) {
+		k_mutex_unlock(&ipuc_mutex);
+		return SUIT_PLAT_ERR_NOT_FOUND;
+	}
+
+	if (ipuc_entry->usage != IPUC_IPC_IN_PLACE_UPDATE ||
+	    ipuc_entry->ipc_client_id != ipc_client_id) {
+		k_mutex_unlock(&ipuc_mutex);
+		return SUIT_PLAT_ERR_INCORRECT_STATE;
+	}
+
+	if (ipuc_entry->last_chunk_stored) {
+		k_mutex_unlock(&ipuc_mutex);
+		return SUIT_PLAT_ERR_INCORRECT_STATE;
+	}
+
+	if (buffer && chunk_size > 0) {
+
+		struct stream_sink ipuc_sink;
+		bool ipuc_sink_allocated = false;
+
+		if (suit_flash_sink_is_address_supported((uint8_t *)slot_address)) {
+			if (suit_flash_sink_get(&ipuc_sink, (uint8_t *)slot_address, slot_size) ==
+			    SUIT_PLAT_SUCCESS) {
+				ipuc_sink_allocated = true;
+			}
+		}
+
+		if (!ipuc_sink_allocated) {
+			LOG_ERR("IPUC sink not allocated");
+			k_mutex_unlock(&ipuc_mutex);
+			return SUIT_PLAT_ERR_IO;
+		}
+
+		if (ipuc_sink.seek == NULL ||
+		    ipuc_sink.seek(ipuc_sink.ctx, offset) != SUIT_PLAT_SUCCESS) {
+			LOG_ERR("Seek operation not supported or failed");
+			if (ipuc_sink.release) {
+				ipuc_sink.release(ipuc_sink.ctx);
+			}
+			k_mutex_unlock(&ipuc_mutex);
+			return SUIT_PLAT_ERR_IO;
+		}
+
+		if (suit_generic_address_streamer_stream((uint8_t *)buffer, chunk_size,
```

Comment:
If I understand correctly `buffer` is simply a RAM buffer - please correct me if I am wrong.
If that is the case, then simply using `ipuc_sink.write` would be enough, `suit_generic_address_streamer_stream` would be a unnecessary additional code executing.